Adds tutorial for creating a worker

desertaxle opened the pull request:
<!-- 
Thanks for opening a pull request to Prefect! We've got a few requests to help us review contributions:

- Make sure that your title neatly summarizes the proposed changes.
- Provide a short overview of the change and the value it adds.
- Share an example to help us understand the change in user experience.
- Confirm that you've done common tasks so we can give a timely review.

Happy engineering!
-->

<!-- Include an overview here -->
Adds an advanced for creating a worker. Will be used as a reference for creating new worker types for major cloud providers, and will incorporate feedback from the developers using it prior to merging.

This appears to be the first tutorial related to extending Prefect through the use of Prefect primitives and concepts, so I'm not sure it fits best under 'Tutorials`, but I'm open to moving it elsewhere.

### Example
<!-- 
Share an example of the change in action.

A code blurb is best. Changes to features should include an example that is executable by a new user.
If changing documentation, a link to a preview of the page is great.
 -->

### Checklist
<!-- These boxes may be checked after opening the pull request. -->

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [ ] This pull request includes tests or only affects documentation.
- [ ] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.
  <!-- If you do not have permission to add a label, a maintainer will add one for you -->

cicdw commented on docs/tutorials/creating-a-worker.md:
```diff
@@ -0,0 +1,387 @@
+
+Prefect sets values for each attribute before giving the configuration to the worker. If you want to customize the values of these attributes, use the [`prepare_for_flow_run`](/api-ref/prefect/workers/base/#prefect.workers.base.BaseJobConfiguration.prepare_for_flow_run) method.
+
+Here's an example `prepare_for_flow_run` method that sets add a label to the execution environment:
```
```suggestion
Here's an example `prepare_for_flow_run` method that adds a label to the execution environment:
```

cicdw reviewed (commented):
a few comments as a first pass - this makes me realize that there is an intermediate level tutorial we might consider for customizing the job template for specifically a k8s worker; i'm guessing that will be one of the more popular requests

desertaxle commented on docs/tutorials/creating-a-worker.md:
```diff
@@ -0,0 +1,387 @@
+
+A worker developer defines the configuration their worker needs to function with a class extending [`BaseJobConfiguration`](/api-ref/prefect/workers/base/#prefect.workers.base.BaseJobConfiguration). 
+
+`BaseJobConfiguration` has attributes that are common to all workers: 
```
`type` isn't inferred from the class name and needs to be explicitly set right now.

serinamarie commented on src/prefect/workers/process.py:
```diff
@@ -1,3 +1,22 @@
+"""
+<span class="badge-api beta"/>
+
+Module containing the Docker worker used for executing flow runs as subprocesses.
```
```suggestion
Module containing the Process worker used for executing flow runs as subprocesses.
```

serinamarie commented on docs/tutorials/creating-a-worker.md:
````diff
@@ -0,0 +1,387 @@
+        ...
+```
+
+The `run` method is passed the flow run to execute, the execution environment configuration for the flow run, and a task status object that allows the worker to track whether the flow run was submitted successfully.
````
```suggestion
The `run` method is passed: the flow run to execute, the execution environment configuration for the flow run, and a task status object that allows the worker to track whether the flow run was submitted successfully.
```

serinamarie commented on docs/tutorials/creating-a-worker.md:
```diff
@@ -0,0 +1,387 @@
+
+### Worker Implementation Example
+
+Below is an example of a worker implementation. This example is not intended to be a complete implementation but to illustrate the abovementioned concepts.
```
```suggestion
Below is an example of a worker implementation. This example is not intended to be a complete implementation but to illustrate the aforementioned concepts.
```

serinamarie commented on docs/tutorials/creating-a-worker.md:
````diff
@@ -0,0 +1,387 @@
+        )
+```
+
+Most of the execution logic is omitted from the example above, but it shows that the typical flow for execution in the `run` method is:
````
```suggestion
Most of the execution logic is omitted from the example above, but it shows that the typical flow run for execution in the `run` method is:
```

serinamarie commented on docs/tutorials/creating-a-worker.md:
````diff
@@ -0,0 +1,387 @@
+```python
+entry_points={
+    "prefect.collections": [
+        "my_pacakge_name = my_worker_module",
````
```suggestion
        "my_package_name = my_worker_module",
```

serinamarie reviewed (commented):
Looks nice so far :)

serinamarie commented on docs/tutorials/creating-a-worker.md:
```diff
@@ -0,0 +1,387 @@
+    - recipes
+---
+
+# Creating a Worker
```
Any thoughts on this heading vs "Developing a worker"? I'm wondering if users who just want to start a worker or learn more about workers might end up here accidentally.

billpalombi replied:
+1 this title is a bit too vague. What about "Developing a new worker type"?

billpalombi commented on docs/concepts/work-pools.md:
```diff
@@ -324,6 +324,16 @@ Workers are similar to agents, but offer greater control over infrastructure con
+| Worker Type | Description | Required Package |
+| --- | --- | --- |
+| [`process`](/api-ref/prefect/workers/process/) | Executes flow runs in subprocesses. | |
+| [`kubernetes`](https://prefecthq.github.io/prefect-kubernetes/worker/) | Executes flow runs within Kubernetes jobs. | `prefect-kubernetes` |
```
Should this be _as_ Kubernetes jobs instead of _within_ Kubernetes jobs?

desertaxle replied:
```suggestion
| [`kubernetes`](https://prefecthq.github.io/prefect-kubernetes/worker/) | Executes flow runs as Kubernetes jobs. | `prefect-kubernetes` |
```

billpalombi commented on docs/tutorials/developing-a-new-worker-type.md:
```diff
@@ -0,0 +1,387 @@
+    This tutorial is for users who want to extend the Prefect framework and completing this successfully will require deep knowledge of Prefect concepts. For standard use cases, we recommend using one of the [available workers](/concepts/work-pools/#worker-types) instead.
+
+
+Prefect workers are responsible for creating execution environments and starting flow runs within those execution environments. 
```
```suggestion
Prefect workers are responsible for setting up execution infrastructure and starting flow runs on that infrastructure.
```

billpalombi commented on docs/tutorials/developing-a-new-worker-type.md:
```diff
@@ -0,0 +1,387 @@
+
+Prefect workers are responsible for creating execution environments and starting flow runs within those execution environments. 
+
+A list of available workers can be found in the [Work Pools, Workers & Agents documentation](/concepts/work-pools/#worker-types). What if you want to execute your flow runs on infrastructure that doesn't have a worker available? This tutorial will walk you through creating a custom worker that can run your flows on your chosen infrastructure.
```
```suggestion
A list of available workers can be found in the [Work Pools, Workers & Agents documentation](/concepts/work-pools/#worker-types). What if you want to execute your flow runs on infrastructure that doesn't have an available worker type? This tutorial will walk you through creating a custom worker that can run your flows on your chosen infrastructure.
```

billpalombi commented on docs/tutorials/developing-a-new-worker-type.md:
```diff
@@ -0,0 +1,387 @@
+When setting up an execution environment for a flow run, a worker receives configuration for the infrastructure it is designed to work with. Examples of configuration values include memory allocation, CPU allocation, credentials, image name, etc. The worker then uses this configuration to create the execution environment and start the flow run.
+
+!!! tip "How are the configuration values populated?"
+    The work pool that a worker polls for flow runs has a [base job template](/concepts/work-pools/#base-job-template) associated with it which is the contract for how configuration values populate for each flow run.
```
```suggestion
    The work pool that a worker polls for flow runs has a [base job template](/concepts/work-pools/#base-job-template) associated with it. The template is the contract for how configuration values populate for each flow run.
```

billpalombi commented on docs/tutorials/developing-a-new-worker-type.md:
```diff
@@ -0,0 +1,387 @@
+
+This base job template defines what values can be provided by deployment creators on a per-deployment basis and how those provided values will be translated into the configuration values that the worker will use to create the execution environment.
+
+Notice that each attribute for the class was added in the `job_configuration` section with placeholders whose name matches the attribute name. The `variables` section was also populated with the OpenAPI schema for each attribute. If a configuration class is used without explicitly declaring any template variables, then the template variables will be inferred from the configuration class attribute ...
```
```suggestion
Notice that each attribute for the class was added in the `job_configuration` section with placeholders whose name matches the attribute name. The `variables` section was also populated with the OpenAPI schema for each attribute. If a configuration class is used without explicitly declaring any template variables, the template variables will be inferred from the configuration class attributes.
```

billpalombi commented on docs/tutorials/developing-a-new-worker-type.md:
```diff
@@ -0,0 +1,387 @@
+
+## Worker Implementation
+
+Workers handle creating execution environments using provided configuration. Workers also observe the execution environment as the flow run executes and report any crashes to the Prefect API.
```
```suggestion
Workers set up execution environments using provided configuration. Workers also observe the execution environment as the flow run executes and report any crashes to the Prefect API.
```

billpalombi reviewed (approved):
Minor nits - otherwise lgtm

desertaxle commented:
I added a section for adding flow run cancelation support to a worker and incorporated feedback from our recent round of worker development. I think it's worthwhile to give this tutorial one final review before merging.

serinamarie commented on docs/tutorials/developing-a-new-worker-type.md:
```diff
@@ -165,6 +165,40 @@ job_configuration:
+
+These rules allow worker developers and work pool maintainers to define template variables that can be complex types like dictionaries and lists. These rules also mean that worker developers should give reasonable default values to job configuration fields whenever possible because values are not guaranteed to be provided if template variables are unset.
+
+### Template Variable Usage Strategies
```
🔥

serinamarie reviewed (approved):
Looks good to me! Although it may be best to have someone who developed a worker review this.

serinamarie commented on docs/tutorials/developing-a-new-worker-type.md:
```diff
@@ -0,0 +1,444 @@
+
+These rules allow worker developers and work pool maintainers to define template variables that can be complex types like dictionaries and lists. These rules also mean that worker developers should give reasonable default values to job configuration fields whenever possible because values are not guaranteed to be provided if template variables are unset.
+
+### Template Variable Usage Strategies
+
+Template variables define the interface that deployment creators interact with to configure the execution environments of their deployments. The complexity of this interface can be controlled via the template variables that are defined for a base job template. This control allows work pool maintainers to find a point along the spectrum of flexibility and simplicity appropriate for their organizat ...
+
+There are two patterns that are represented in current worker implementations:
```
🔥

jakekaplan commented on mkdocs.yml:
```diff
@@ -120,6 +120,7 @@ nav:
         - Projects: tutorials/projects.md
         - Deployments: tutorials/deployments.md
         - Storage & Infrastructure: tutorials/storage.md
+        - Developing a New Worker Type: tutorials/developing-a-new-worker-type.md
```
I think this got litigated already but I could see this under `Guides` as opposed to tutorials

jakekaplan commented on docs/tutorials/developing-a-new-worker-type.md:
```diff
@@ -0,0 +1,444 @@
+
+In the IaC pattern, it's often useful to use template variables to template portions of the declarative syntax which then can be used to generate the declarative syntax into a final form. 
+
+This approach allows work pool creators to provide a simpler interface to deployment creators while also controlling which portions of infrastructure are configurable by deployment creators.
```
👏

jakekaplan reviewed (approved):
looks great to me! thanks for adding the extra variables section

desertaxle commented on mkdocs.yml:
```diff
@@ -120,6 +120,7 @@ nav:
         - Projects: tutorials/projects.md
         - Deployments: tutorials/deployments.md
         - Storage & Infrastructure: tutorials/storage.md
+        - Developing a New Worker Type: tutorials/developing-a-new-worker-type.md
```
Yeah, I think the consensus was to have it in tutorials for now, but I also expect that it will move in the future as we produce more content like this.
